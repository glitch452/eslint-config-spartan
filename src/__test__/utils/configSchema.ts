import { z } from 'zod';

const INITIAL_ECMA_VERSION = 3;
const INITIAL_ECMA_YEAR = 2015;
const NEXT_YEAR = new Date().getFullYear() + 1;
const MIN_ECMA_VERSION = 5;
const MAX_ECMA_VERSION = NEXT_YEAR - INITIAL_ECMA_YEAR + MIN_ECMA_VERSION + 1;

const ecmaVersionSchema = z.union([
  z.literal('latest'),
  z.literal(INITIAL_ECMA_VERSION),
  z.number().int().min(MIN_ECMA_VERSION).max(MAX_ECMA_VERSION),
  z.number().int().min(INITIAL_ECMA_YEAR).max(NEXT_YEAR),
]);
const sourceTypeSchema = z.enum(['script', 'module', 'commonjs']);
const globalConfSchema = z.boolean().or(z.enum(['off', 'readable', 'readonly', 'writable', 'writeable']));

const MIN_SEVERITY = 0;
const MAX_SEVERITY = 2;
const severitySchema = z.number().int().min(MIN_SEVERITY).max(MAX_SEVERITY);
const stringSeveritySchema = z.enum(['off', 'warn', 'error']);

const objectMetaPropertiesSchema = z.object({
  /** @deprecated Use `meta.name` instead. */
  name: z.string().optional(),
  /** @deprecated Use `meta.version` instead. */
  version: z.string().optional(),
  meta: z.object({ name: z.string().optional(), version: z.string().optional() }).optional(),
});

const esTreeParserSchema = z.union([
  objectMetaPropertiesSchema.extend({ parse: z.function() }),
  objectMetaPropertiesSchema.extend({ parseForESLint: z.function() }),
]);

const processorSchema = objectMetaPropertiesSchema.extend({
  supportsAutofix: z.boolean().optional(),
  preprocess: z.function(),
  postprocess: z.function(),
});

const parserOptionsSchema = z
  .object({
    ecmaVersion: ecmaVersionSchema,
    sourceType: sourceTypeSchema,
    ecmaFeatures: z
      .object({
        globalReturn: z.boolean(),
        impliedStrict: z.boolean(),
        jsx: z.boolean(),
        experimentalObjectRestSpread: z.boolean(),
      })
      .partial()
      .catchall(z.any()),
  })
  .partial()
  .catchall(z.any());

const languageOptionsSchema = z
  .object({
    ecmaVersion: ecmaVersionSchema,
    sourceType: sourceTypeSchema,
    globals: z.record(globalConfSchema),
    parser: esTreeParserSchema,
    parserOptions: parserOptionsSchema,
  })
  .partial();

const linterOptionsSchema = z
  .object({
    noInlineConfig: z.boolean().optional(),
    reportUnusedDisableDirectives: z.union([severitySchema, stringSeveritySchema, z.boolean()]).optional(),
  })
  .optional();

const externalSpecifierSchema = z.object({
  name: z.string().optional(),
  url: z.string().optional(),
});

const replacedBySchema = z.object({
  message: z.string().optional(),
  url: z.string().optional(),
  plugin: externalSpecifierSchema.optional(),
  rule: externalSpecifierSchema.optional(),
});

const deprecatedInfoSchema = z.object({
  message: z.string().optional(),
  url: z.string().optional(),
  replacedBy: replacedBySchema.array().optional(),
  deprecatedSince: z.string().optional(),
  availableUntil: z.string().nullable().optional(),
});

const ruleMetaDataSchema = z
  .object({
    docs: z
      .object({
        description: z.string(),
        category: z.string(),
        // Added z.string() to accommodate storybook plugin
        // Added z.record(z.unknown()) to accommodate typescript-eslint plugin
        recommended: z.union([z.boolean(), z.string(), z.record(z.unknown())]),
        url: z.string(),
      })
      .partial(),
    messages: z.record(z.string()),
    // Added z.null() added to accommodate the cypress plugin
    fixable: z.enum(['code', 'whitespace']).nullable(),
    schema: z.unknown(),
    deprecated: z.boolean().or(deprecatedInfoSchema),
    replacedBy: z.string().array().readonly(),
    // Added 'error' to accommodate the security plugin
    type: z.enum(['problem', 'suggestion', 'layout', 'error']),
    hasSuggestions: z.boolean(),
  })
  .partial();

const ruleModuleSchema = z.object({
  create: z.function(),
  meta: ruleMetaDataSchema.optional(),
});

const pluginSchema = z
  .object({
    configs: z.record(z.unknown()),
    environments: z.record(
      z.object({ globals: z.record(globalConfSchema), parserOptions: parserOptionsSchema }).partial(),
    ),
    processors: z.record(processorSchema),
    rules: z.record(ruleModuleSchema),
  })
  .partial();

export const configSchema = z
  .object({
    name: z.string(),
    files: z.array(z.string().or(z.string().array())),
    ignores: z.string().array(),
    languageOptions: languageOptionsSchema,
    linterOptions: linterOptionsSchema,
    processor: z.string().or(processorSchema),
    plugins: z.record(pluginSchema),
    rules: z.unknown(),
    settings: z.record(z.unknown()),
  })
  .partial();
