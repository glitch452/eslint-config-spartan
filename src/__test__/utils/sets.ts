type Widen<T> = T | (TSReset.WidenLiteral<T> & {});

/**
 * `A âˆ© B`.
 * Compute the intersection of unique elements contained in both {@link a} and {@link b}.
 * @param a
 * @param b
 * @returns
 */
export function intersection<A, B>(a: Widen<A>[], b: Widen<B>[]): Extract<Widen<A>, Widen<B>>[] {
  const setB = new Set<Widen<A> | Widen<B>>(b);
  return [...new Set<Widen<A>>(a)].filter((x): x is Extract<Widen<A>, Widen<B>> => setB.has(x));
}

/**
 * `A - B` a.k.a. `A / B`.
 * Compute the unique elements in {@link a} that do not also exist in {@link b}.
 * @param a
 * @param b
 * @returns
 */
export function difference<A, B>(a: A[], b: B[]): Widen<A>[] {
  const setB = new Set<A | B>(b);
  return [...new Set<A>(a)].filter((x): x is A => !setB.has(x));
}
