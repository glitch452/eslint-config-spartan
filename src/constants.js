export const configNamePrefix = 'eslint-config-spartan';

export const prefixes = /** @type {const} */ ({
  angular: '@angular-eslint',
  angularTemplate: '@angular-eslint/template',
  chaiExpect: 'chai-expect',
  chaiFriendly: 'chai-friendly',
  cypress: 'cypress',
  i18next: 'i18next',
  i18nJson: 'i18n-json',
  import: 'import',
  jest: 'jest',
  jestDom: 'jest-dom',
  jsDoc: 'jsdoc',
  jsxA11y: 'jsx-a11y',
  mdx: 'mdx',
  mocha: 'mocha',
  next: '@next/next',
  playwright: 'playwright',
  promise: 'promise',
  react: 'react',
  reactHooks: 'react-hooks',
  regExp: 'regexp',
  security: 'security',
  storybook: 'storybook',
  stylistic: '@stylistic',
  tailwindCss: 'tailwindcss',
  tanStackQuery: '@tanstack/query',
  testingLibrary: 'testing-library',
  typescriptEnum: 'typescript-enum',
  typescriptEsLint: '@typescript-eslint',
  unicorn: 'unicorn',
  unusedImports: 'unused-imports',
  vitest: 'vitest',
});
