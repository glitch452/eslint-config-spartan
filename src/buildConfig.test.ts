import typescriptEsLint from 'typescript-eslint';
import stylisticPlugin from '@stylistic/eslint-plugin';
import importPlugin from 'eslint-plugin-import';
import securityPlugin from 'eslint-plugin-security';
import unusedImportsPlugin from 'eslint-plugin-unused-imports';
import typescriptEnumPlugin from 'eslint-plugin-typescript-enum';
import unicornPlugin from 'eslint-plugin-unicorn';
import { configSchema } from './__test__/utils/configSchema.js';
import { difference, intersection } from './__test__/utils/sets.js';
import { files } from './utils/index.js';
import { CONFIG_NAME_PREFIX, prefixes } from './constants.js';
import {
  getDeprecatedRules,
  getEnabledRules,
  getTypeInfoRequiredRules,
  getWarnRules,
  listRules,
} from './__test__/utils/rules.js';
import { buildConfig } from './buildConfig.js';
import { Linter } from 'eslint';

describe(buildConfig.name, () => {
  const configs = buildConfig();

  const expectedFilesGeneral = [files.jsTs];

  const testTable = [
    { id: 'base/eslint/recommended', expectedFiles: expectedFilesGeneral },
    { id: 'base/typescript-eslint/base', expectedFiles: expectedFilesGeneral },
    { id: 'base/typescript-eslint/eslint-recommended', expectedFiles: expectedFilesGeneral },
    { id: 'base/typescript-eslint/recommended', expectedFiles: expectedFilesGeneral },
    { id: 'base', expectedFiles: expectedFilesGeneral },
    { id: 'base/commonJs', expectedFiles: [files.cjs] },
  ] as const;

  const validRules = [
    ...new Linter({ configType: 'eslintrc' }).getRules().keys(),
    ...listRules(typescriptEsLint.plugin.rules as any, prefixes.typescriptEsLint),
    ...listRules(stylisticPlugin.rules as any, prefixes.stylistic),
    ...listRules(importPlugin.rules, prefixes.import),
    ...listRules(securityPlugin.rules, prefixes.security),
    ...listRules(unusedImportsPlugin.rules, prefixes.unusedImports),
    ...listRules(typescriptEnumPlugin.rules, prefixes.typescriptEnum),
    ...listRules(unicornPlugin.rules, prefixes.unicorn),
  ];
  const deprecatedRules = [
    ...getDeprecatedRules(new Linter({ configType: 'eslintrc' }).getRules()),
    ...getDeprecatedRules(typescriptEsLint.plugin.rules as any, prefixes.typescriptEsLint),
    ...getDeprecatedRules(importPlugin.rules, prefixes.import),
    ...getDeprecatedRules(securityPlugin.rules, prefixes.security),
    ...getDeprecatedRules(unusedImportsPlugin.rules, prefixes.unusedImports),
    ...getDeprecatedRules(typescriptEnumPlugin.rules, prefixes.typescriptEnum),
  ];
  const typeInfoRequiredRules = getTypeInfoRequiredRules(
    typescriptEsLint.plugin.rules as any,
    prefixes.typescriptEsLint,
  );

  it('should test all the returned configs', () => {
    expect(testTable).toHaveLength(configs.length);
  });

  describe.each(testTable)('$id', (item) => {
    const { id, expectedFiles } = item;
    const index = testTable.indexOf(item);
    const config = configs[index];

    it('should create a valid eslint config', () => {
      const actual = () => configSchema.parse(config);
      expect(actual).not.toThrow();
    });

    it('should configure the expected rules', () => {
      expect(config.rules).toMatchSnapshot();
    });

    it('should set the name value', () => {
      expect(config.name).toBe(`${CONFIG_NAME_PREFIX}/${id}`);
    });

    it('should set the default files value', () => {
      expect(config.files).toStrictEqual(expectedFiles);
    });

    it('should only configure rules that exist', () => {
      const configuredRules = Object.keys(config.rules ?? {});
      const actual = difference(configuredRules, validRules);
      expect(actual).toStrictEqual([]);
    });

    it('should not configure rules that are deprecated', () => {
      const configuredAndEnabledRules = getEnabledRules(config.rules);
      const actual = intersection(configuredAndEnabledRules, deprecatedRules);
      expect(actual).toStrictEqual([]);
    });

    it('should not configure any rules as warn', () => {
      const configuredAndWarnRules = getWarnRules(config.rules);
      expect(configuredAndWarnRules).toStrictEqual([]);
    });

    it('should not configure rules that require type information', () => {
      const configuredTsEsLintRules = Object.keys(config.rules ?? {}).filter((x) =>
        x.startsWith(prefixes.typescriptEsLint),
      );
      const actual = intersection(configuredTsEsLintRules, typeInfoRequiredRules);
      expect(actual).toStrictEqual([]);
    });
  });
});
