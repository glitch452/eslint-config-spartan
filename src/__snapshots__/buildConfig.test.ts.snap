// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildConfig > 'base' > should configure the expected rules 1`] = `
{
  "@stylistic/array-bracket-spacing": [
    "error",
    "never",
  ],
  "@stylistic/arrow-parens": [
    "error",
    "always",
  ],
  "@stylistic/arrow-spacing": [
    "error",
    {
      "after": true,
      "before": true,
    },
  ],
  "@stylistic/block-spacing": [
    "error",
    "always",
  ],
  "@stylistic/brace-style": [
    "error",
    "1tbs",
    {
      "allowSingleLine": false,
    },
  ],
  "@stylistic/comma-dangle": [
    "error",
    "always-multiline",
  ],
  "@stylistic/comma-spacing": [
    "error",
    {
      "after": true,
      "before": false,
    },
  ],
  "@stylistic/comma-style": [
    "error",
    "last",
  ],
  "@stylistic/computed-property-spacing": [
    "error",
    "never",
    {
      "enforceForClassMembers": true,
    },
  ],
  "@stylistic/dot-location": [
    "error",
    "property",
  ],
  "@stylistic/eol-last": "error",
  "@stylistic/generator-star-spacing": [
    "error",
    {
      "after": true,
      "before": false,
    },
  ],
  "@stylistic/indent": [
    "error",
    2,
    {
      "ArrayExpression": 1,
      "CallExpression": {
        "arguments": 1,
      },
      "FunctionDeclaration": {
        "body": 1,
        "parameters": 1,
      },
      "FunctionExpression": {
        "body": 1,
        "parameters": 1,
      },
      "ImportDeclaration": 1,
      "MemberExpression": 1,
      "ObjectExpression": 1,
      "SwitchCase": 1,
      "VariableDeclarator": 1,
      "flatTernaryExpressions": false,
      "ignoreComments": false,
      "ignoredNodes": [
        "TSUnionType",
        "TSIntersectionType",
        "TSTypeParameterInstantiation",
        "FunctionExpression > .params[decorators.length > 0]",
        "FunctionExpression > .params > :matches(Decorator, :not(:first-child))",
      ],
      "offsetTernaryExpressions": true,
      "outerIIFEBody": 1,
      "tabLength": 2,
    },
  ],
  "@stylistic/indent-binary-ops": [
    "error",
    2,
  ],
  "@stylistic/jsx-closing-bracket-location": "error",
  "@stylistic/jsx-closing-tag-location": "error",
  "@stylistic/jsx-curly-brace-presence": [
    "error",
    {
      "propElementValues": "always",
    },
  ],
  "@stylistic/jsx-curly-newline": "error",
  "@stylistic/jsx-curly-spacing": [
    "error",
    {
      "children": true,
      "when": "never",
    },
  ],
  "@stylistic/jsx-equals-spacing": "error",
  "@stylistic/jsx-first-prop-new-line": "error",
  "@stylistic/jsx-function-call-newline": [
    "error",
    "multiline",
  ],
  "@stylistic/jsx-indent-props": [
    "error",
    2,
  ],
  "@stylistic/jsx-max-props-per-line": [
    "error",
    {
      "maximum": 1,
      "when": "multiline",
    },
  ],
  "@stylistic/jsx-one-expression-per-line": [
    "error",
    {
      "allow": "single-child",
    },
  ],
  "@stylistic/jsx-quotes": "error",
  "@stylistic/jsx-tag-spacing": [
    "error",
    {
      "afterOpening": "never",
      "beforeClosing": "never",
      "beforeSelfClosing": "always",
      "closingSlash": "never",
    },
  ],
  "@stylistic/jsx-wrap-multilines": [
    "error",
    {
      "arrow": "parens-new-line",
      "assignment": "parens-new-line",
      "condition": "parens-new-line",
      "declaration": "parens-new-line",
      "logical": "parens-new-line",
      "prop": "parens-new-line",
      "propertyValue": "parens-new-line",
      "return": "parens-new-line",
    },
  ],
  "@stylistic/key-spacing": [
    "error",
    {
      "afterColon": true,
      "beforeColon": false,
    },
  ],
  "@stylistic/keyword-spacing": [
    "error",
    {
      "after": true,
      "before": true,
    },
  ],
  "@stylistic/lines-between-class-members": [
    "error",
    "always",
    {
      "exceptAfterSingleLine": true,
    },
  ],
  "@stylistic/max-statements-per-line": [
    "error",
    {
      "max": 1,
    },
  ],
  "@stylistic/member-delimiter-style": "error",
  "@stylistic/multiline-ternary": [
    "error",
    "always-multiline",
  ],
  "@stylistic/new-parens": "error",
  "@stylistic/no-extra-parens": [
    "error",
    "functions",
  ],
  "@stylistic/no-floating-decimal": "error",
  "@stylistic/no-mixed-operators": [
    "error",
    {
      "allowSamePrecedence": true,
      "groups": [
        [
          "==",
          "!=",
          "===",
          "!==",
          ">",
          ">=",
          "<",
          "<=",
        ],
        [
          "&&",
          "||",
        ],
        [
          "in",
          "instanceof",
        ],
      ],
    },
  ],
  "@stylistic/no-mixed-spaces-and-tabs": "error",
  "@stylistic/no-multi-spaces": "error",
  "@stylistic/no-multiple-empty-lines": [
    "error",
    {
      "max": 1,
      "maxBOF": 0,
      "maxEOF": 0,
    },
  ],
  "@stylistic/no-tabs": "error",
  "@stylistic/no-trailing-spaces": "error",
  "@stylistic/no-whitespace-before-property": "error",
  "@stylistic/object-curly-spacing": [
    "error",
    "always",
  ],
  "@stylistic/operator-linebreak": [
    "error",
    "before",
  ],
  "@stylistic/padded-blocks": [
    "error",
    {
      "blocks": "never",
      "classes": "never",
      "switches": "never",
    },
  ],
  "@stylistic/quote-props": [
    "error",
    "consistent-as-needed",
  ],
  "@stylistic/quotes": [
    "error",
    "single",
    {
      "allowTemplateLiterals": true,
      "avoidEscape": false,
    },
  ],
  "@stylistic/rest-spread-spacing": [
    "error",
    "never",
  ],
  "@stylistic/semi": [
    "error",
    "always",
  ],
  "@stylistic/semi-spacing": [
    "error",
    {
      "after": true,
      "before": false,
    },
  ],
  "@stylistic/space-before-blocks": [
    "error",
    "always",
  ],
  "@stylistic/space-before-function-paren": [
    "error",
    {
      "anonymous": "always",
      "asyncArrow": "always",
      "named": "never",
    },
  ],
  "@stylistic/space-in-parens": [
    "error",
    "never",
  ],
  "@stylistic/space-infix-ops": "error",
  "@stylistic/space-unary-ops": [
    "error",
    {
      "nonwords": false,
      "words": true,
    },
  ],
  "@stylistic/spaced-comment": [
    "error",
    "always",
    {
      "block": {
        "balanced": true,
        "exceptions": [
          "*",
        ],
        "markers": [
          "!",
        ],
      },
      "line": {
        "exceptions": [
          "/",
          "#",
        ],
        "markers": [
          "/",
        ],
      },
    },
  ],
  "@stylistic/template-curly-spacing": "error",
  "@stylistic/template-tag-spacing": [
    "error",
    "never",
  ],
  "@stylistic/type-annotation-spacing": [
    "error",
    {},
  ],
  "@stylistic/type-generic-spacing": "error",
  "@stylistic/type-named-tuple-spacing": "error",
  "@stylistic/wrap-iife": [
    "error",
    "any",
    {
      "functionPrototypeMethods": true,
    },
  ],
  "@stylistic/yield-star-spacing": [
    "error",
    {
      "after": true,
      "before": false,
    },
  ],
  "@typescript-eslint/adjacent-overload-signatures": "error",
  "@typescript-eslint/consistent-type-assertions": [
    "error",
    {
      "assertionStyle": "never",
    },
  ],
  "@typescript-eslint/consistent-type-definitions": [
    "error",
    "interface",
  ],
  "@typescript-eslint/default-param-last": "error",
  "@typescript-eslint/explicit-member-accessibility": [
    "error",
    {
      "accessibility": "no-public",
    },
  ],
  "@typescript-eslint/explicit-module-boundary-types": "off",
  "@typescript-eslint/method-signature-style": [
    "error",
    "property",
  ],
  "@typescript-eslint/no-confusing-non-null-assertion": "error",
  "@typescript-eslint/no-dupe-class-members": "error",
  "@typescript-eslint/no-dynamic-delete": "error",
  "@typescript-eslint/no-inferrable-types": [
    "error",
    {
      "ignoreParameters": true,
    },
  ],
  "@typescript-eslint/no-invalid-void-type": [
    "error",
    {
      "allowAsThisParameter": true,
    },
  ],
  "@typescript-eslint/no-magic-numbers": [
    "error",
    {
      "ignore": [
        -1,
        0,
        1,
      ],
      "ignoreEnums": true,
      "ignoreNumericLiteralTypes": false,
      "ignoreReadonlyClassProperties": true,
    },
  ],
  "@typescript-eslint/no-non-null-assertion": "error",
  "@typescript-eslint/no-require-imports": "error",
  "@typescript-eslint/no-shadow": "error",
  "@typescript-eslint/no-unused-vars": "off",
  "@typescript-eslint/no-useless-constructor": "error",
  "@typescript-eslint/prefer-for-of": "error",
  "array-callback-return": "error",
  "curly": "error",
  "default-case": [
    "error",
    {
      "commentPattern": "^skip\\sdefault",
    },
  ],
  "default-case-last": "error",
  "default-param-last": "off",
  "dot-notation": "error",
  "eqeqeq": "error",
  "import/default": "error",
  "import/export": "error",
  "import/first": "error",
  "import/named": "error",
  "import/namespace": "error",
  "import/newline-after-import": "error",
  "import/no-absolute-path": "error",
  "import/no-cycle": "error",
  "import/no-deprecated": "error",
  "import/no-duplicates": [
    "error",
    {
      "considerQueryString": true,
    },
  ],
  "import/no-empty-named-blocks": "error",
  "import/no-import-module-exports": "error",
  "import/no-named-as-default": "error",
  "import/no-named-as-default-member": "off",
  "import/no-self-import": "error",
  "import/no-unresolved": "off",
  "import/no-useless-path-segments": "error",
  "no-alert": "error",
  "no-caller": "error",
  "no-constructor-return": "error",
  "no-dupe-class-members": "off",
  "no-duplicate-imports": "off",
  "no-else-return": [
    "error",
    {
      "allowElseIf": false,
    },
  ],
  "no-eq-null": "error",
  "no-eval": "error",
  "no-extra-bind": "error",
  "no-implied-eval": "error",
  "no-iterator": "error",
  "no-labels": "error",
  "no-magic-numbers": "off",
  "no-new-func": "error",
  "no-param-reassign": "error",
  "no-proto": "error",
  "no-return-assign": "error",
  "no-script-url": "error",
  "no-self-compare": "error",
  "no-sequences": "error",
  "no-shadow": "off",
  "no-throw-literal": "error",
  "no-unassigned-vars": "error",
  "no-unmodified-loop-condition": "error",
  "no-unused-vars": "off",
  "no-useless-call": "error",
  "no-useless-computed-key": "error",
  "no-useless-concat": "error",
  "no-useless-rename": "error",
  "no-useless-return": "error",
  "object-shorthand": [
    "error",
    "always",
  ],
  "prefer-template": "error",
  "require-atomic-updates": "error",
  "security/detect-bidi-characters": "error",
  "security/detect-buffer-noassert": "error",
  "security/detect-child-process": "error",
  "security/detect-disable-mustache-escape": "error",
  "security/detect-eval-with-expression": "error",
  "security/detect-new-buffer": "error",
  "security/detect-no-csrf-before-method-override": "error",
  "security/detect-non-literal-fs-filename": "error",
  "security/detect-non-literal-regexp": "error",
  "security/detect-non-literal-require": "error",
  "security/detect-object-injection": "off",
  "security/detect-possible-timing-attacks": "error",
  "security/detect-pseudoRandomBytes": "error",
  "security/detect-unsafe-regex": "error",
  "sort-imports": [
    "error",
    {
      "ignoreDeclarationSort": true,
    },
  ],
  "typescript-enum/no-const-enum": "error",
  "typescript-enum/no-enum": "error",
  "unused-imports/no-unused-imports": "error",
  "unused-imports/no-unused-vars": [
    "error",
    {
      "args": "after-used",
      "argsIgnorePattern": "^_",
      "vars": "all",
      "varsIgnorePattern": "^_",
    },
  ],
  "yoda": [
    "error",
    "never",
    {
      "exceptRange": true,
    },
  ],
}
`;

exports[`buildConfig > 'base/commonJs' > should configure the expected rules 1`] = `
{
  "@typescript-eslint/no-require-imports": "off",
  "@typescript-eslint/no-var-requires": "off",
  "unicorn/prefer-module": "off",
}
`;

exports[`buildConfig > 'base/eslint/recommended' > should configure the expected rules 1`] = `
{
  "constructor-super": "error",
  "for-direction": "error",
  "getter-return": "error",
  "no-async-promise-executor": "error",
  "no-case-declarations": "error",
  "no-class-assign": "error",
  "no-compare-neg-zero": "error",
  "no-cond-assign": "error",
  "no-const-assign": "error",
  "no-constant-binary-expression": "error",
  "no-constant-condition": "error",
  "no-control-regex": "error",
  "no-debugger": "error",
  "no-delete-var": "error",
  "no-dupe-args": "error",
  "no-dupe-class-members": "error",
  "no-dupe-else-if": "error",
  "no-dupe-keys": "error",
  "no-duplicate-case": "error",
  "no-empty": "error",
  "no-empty-character-class": "error",
  "no-empty-pattern": "error",
  "no-empty-static-block": "error",
  "no-ex-assign": "error",
  "no-extra-boolean-cast": "error",
  "no-fallthrough": "error",
  "no-func-assign": "error",
  "no-global-assign": "error",
  "no-import-assign": "error",
  "no-invalid-regexp": "error",
  "no-irregular-whitespace": "error",
  "no-loss-of-precision": "error",
  "no-misleading-character-class": "error",
  "no-new-native-nonconstructor": "error",
  "no-nonoctal-decimal-escape": "error",
  "no-obj-calls": "error",
  "no-octal": "error",
  "no-prototype-builtins": "error",
  "no-redeclare": "error",
  "no-regex-spaces": "error",
  "no-self-assign": "error",
  "no-setter-return": "error",
  "no-shadow-restricted-names": "error",
  "no-sparse-arrays": "error",
  "no-this-before-super": "error",
  "no-undef": "error",
  "no-unexpected-multiline": "error",
  "no-unreachable": "error",
  "no-unsafe-finally": "error",
  "no-unsafe-negation": "error",
  "no-unsafe-optional-chaining": "error",
  "no-unused-labels": "error",
  "no-unused-private-class-members": "error",
  "no-unused-vars": "error",
  "no-useless-backreference": "error",
  "no-useless-catch": "error",
  "no-useless-escape": "error",
  "no-with": "error",
  "require-yield": "error",
  "use-isnan": "error",
  "valid-typeof": "error",
}
`;

exports[`buildConfig > 'base/typescript-eslint/base' > should configure the expected rules 1`] = `undefined`;

exports[`buildConfig > 'base/typescript-eslint/eslint-recomme…' > should configure the expected rules 1`] = `
{
  "constructor-super": "off",
  "getter-return": "off",
  "no-class-assign": "off",
  "no-const-assign": "off",
  "no-dupe-args": "off",
  "no-dupe-class-members": "off",
  "no-dupe-keys": "off",
  "no-func-assign": "off",
  "no-import-assign": "off",
  "no-new-native-nonconstructor": "off",
  "no-new-symbol": "off",
  "no-obj-calls": "off",
  "no-redeclare": "off",
  "no-setter-return": "off",
  "no-this-before-super": "off",
  "no-undef": "off",
  "no-unreachable": "off",
  "no-unsafe-negation": "off",
  "no-var": "error",
  "no-with": "off",
  "prefer-const": "error",
  "prefer-rest-params": "error",
  "prefer-spread": "error",
}
`;

exports[`buildConfig > 'base/typescript-eslint/recommended' > should configure the expected rules 1`] = `
{
  "@typescript-eslint/ban-ts-comment": "error",
  "@typescript-eslint/no-array-constructor": "error",
  "@typescript-eslint/no-duplicate-enum-values": "error",
  "@typescript-eslint/no-empty-object-type": "error",
  "@typescript-eslint/no-explicit-any": "error",
  "@typescript-eslint/no-extra-non-null-assertion": "error",
  "@typescript-eslint/no-misused-new": "error",
  "@typescript-eslint/no-namespace": "error",
  "@typescript-eslint/no-non-null-asserted-optional-chain": "error",
  "@typescript-eslint/no-require-imports": "error",
  "@typescript-eslint/no-this-alias": "error",
  "@typescript-eslint/no-unnecessary-type-constraint": "error",
  "@typescript-eslint/no-unsafe-declaration-merging": "error",
  "@typescript-eslint/no-unsafe-function-type": "error",
  "@typescript-eslint/no-unused-expressions": "error",
  "@typescript-eslint/no-unused-vars": "error",
  "@typescript-eslint/no-wrapper-object-types": "error",
  "@typescript-eslint/prefer-as-const": "error",
  "@typescript-eslint/prefer-namespace-keyword": "error",
  "@typescript-eslint/triple-slash-reference": "error",
  "no-array-constructor": "off",
  "no-unused-expressions": "off",
  "no-unused-vars": "off",
}
`;
